Show how to write matplotlib tests with image comparison (close #106)

jacg opened the pull request:
- [ ] Eliminate inconsistencies between Python 3.6 and other versions
- [ ] Ensure that setting of Agg backend works, and that there are no
      crashes on Travis

Added nose to the conda environment, because matplotlib uses it in its
testing infrastructure. [This should be unnecessary in the long run, because of matplotlib's [migration to pytest](https://github.com/matplotlib/matplotlib/issues/6731)]

Added pytest-mpl to the conda environment, because it facilitates
using matplotlib's testing infrastructure in pytest.

Note that the matplotlib backend must be set to Agg *before* impolting
`matplotlib.pyplot` in order to avoid crashes in headless
environments.

It seems that the labels on the y-axis are rendered slightly higher in
Python 3.6 than in Pythons 3.5 and 2.7. Consequently the example test
fails in Python 3.6.

**NOTE: agg backend setting still doesn't seem to work correctly** in
spite of having tried a couple of solutions.

jacg committed:
Show how to write matplotlib tests with image comparison (close #106)

- [ ] Eliminate inconsistencies between Python 3.6 and other versions
- [ ] Ensure that setting of Agg backend works, and that there are no
      crashes on Travis

Added nose to the conda environment, because matplotlib uses it in its
testing infrastructure.

Added pytest-mpl to the conda environment, because it facilitates
using matplotlib's testing infrastructure in pytest.

Note that the matplotlib backend must be set to Agg *before* impolting
`matplotlib.pyplot` in order to avoid crashes in headless
environments.

It seems that the labels on the y-axis are rendered slightly higher in
Python 3.6 than in Pythons 3.5 and 2.7. Consequently the example test
fails in Python 3.6.

**NOTE: agg backend setting still doesn't seem to work correctly** in
spite of having tried a couple of solutions.

jjgomezcadenas reviewed (commented):
Travis is still failing... is this what we expect?

gonzaponte committed:
Add Zrms to table description

gonzaponte committed:
Add Zrms to KrEvent

gonzaponte committed:
Add method to compute the peak's rms

gonzaponte committed:
Check rms value in tests

gonzaponte committed:
Cosmetics

gonzaponte committed:
PEP8-ify spacing

gonzaponte committed:
Use class methods to avoid code repetition

gonzaponte committed:
Add global rms attribute to Peak

gonzaponte committed:
Include Zrms in base city

gonzaponte committed:
Update test file and fixture with new column

gonzaponte committed:
Improve peak composite

So far, the time array could contain multiple
times the same value, which is non-physical.

gonzaponte committed:
Add new normalization option "center"

This is a shortcut for

norm_strategy = "index"
norm_opts = {"value": central_bin}

where central_bin is the arrays' shape tuple halved.

gonzaponte committed:
Cosmetics: reduce excesive spacing in imports

gonzaponte committed:
Cosmetics: fix spacing around infix operators

gonzaponte committed:
Fix typo

gonzaponte committed:
Keep configuration parameters as class attributes

All attributes are also public now.

gonzaponte committed:
Move variable assigment to inner method

gonzaponte committed:
Introduce `interp_opts` as a parameter

gonzaponte committed:
Fix typos

gonzaponte committed:
Introduce better values for default arguments

gonzaponte committed:
Major refactoring of the interpolation section in Correction class

The section of the class devoted to interpolation has been removed
following the principle of not reinventing the wheel. There is
an efficient and very handy function in scipy to do the job. As a
result of that, some extra posibilities arise, such as linear and
cubic interpolations. The original `RectBivariateSpline` interpolation
has been removed as there is no user case.

gonzaponte committed:
Adapt tests to new arguments

gonzaponte committed:
Set explicitly the interpolation option in tests

gonzaponte committed:
Add test for linear interpolation

gonzaponte committed:
Add test for cubic interpolation

The test is not great, but at least it verifies the cubic option is
working properly.

gonzaponte committed:
Declare frequently used interpolation options

This is also meant to serve as a guide to users.

gonzaponte committed:
Treat both scalars and arrays nicely in Correction calls

This commit allows the class to be called with scalars, something we
already had originally, but that was lost during the refactoring. On
top of that, now the output is of the same type as the input, i.e. if
called with arrays, it returns an array, but if called with scalars,
it returns an scalar.

gonzaponte committed:
Add tests calling  with scalar values

This demonstrates that calling with scalar values raises an error. In
a following commit, the code will handle both arrays and scalar values.

jacg committed:
Proposal for new PMaps interface

gonzaponte committed:
Rename tests

gonzaponte committed:
Redefine test

gonzaponte committed:
Add test that demonstrates bug

the "max" normalization option doesn't work with two-dimensional
corrections

gonzaponte committed:
Fix "max" normalization option

The "max" normalization option worked only for 1D corrections and
crashed if more dimensions where present. This was a consequence of
`np.argmax` returning a flat index.

gonzaponte committed:
#398 Fix overflow warning

https://github.com/nextic/IC/pull/398

[ gonzaponte ]

A test was raising an overflow warning continuously due to the use of
a small value within an exponential function. This commit makes
that value a bit bigger so the warning is not raised, but the test
remains the same.

[ approved by jjgomezcadenas ]

This is an easy one. Since @abotas is procrastinating I take the
liberty to approve it

andLaing committed:
Add SiPM spectra city

Addition of city and related functions for the generation of SiPM
photoelectron spectra (binned pdfs).

Additions include:

1) invisible_cities/cities/sipmPDF.py (and corresponding test) which
   reads raw data and generates two types of spectra for each channel
   once converted into pe (zeroed and in pe and zeroed, in pe and
   difference from mau);

2) invisible_cities/io/hist_io.py, a histogram writer to save on
   output data size.

Changes to existing functions include:

+ change to invisible_cities/peak_functions_c.pyx.signal_sipm to
  include calibration option so that all bin data is saved rather than
  just selected data

  + cal=0 (default) is equivalent to unchanged function,
  + cal=1 saves sipmrawf-mean,
  + cal=2 saves sipmrwf-mean-mau

Undid accidental include of some changes from an unrelated
branch.

jjgomezcadenas committed:
#397 Fix normalization option "max" for N-dimensional corrections

https://github.com/nextic/IC/pull/397

[ gonzaponte ]

The normalization option "max" was failing when used in corrections
with more than 1 dimension. This was not detected before because we
didn't have a test for that case (shame on me!). This PR adds that
test, which also demonstrates the bug and fixes it.

[ approved by jjgomezcadenas ]

This is good stuff. It adds a test to demonstrate bug and then a
solution that fixes the bug. Concise and efficient coding, which
starts to be a trademark of the house.

jacg committed:
Add option for strict adherence to requested voxel size

gonzaponte committed:
Add functions to compute weighted mean, var and std

gonzaponte committed:
Add first test to `weighted_mean_and_var`

This test sets all weights to a random value, but common to all data
points. This must be equivalent to unweighted data.

gonzaponte committed:
Add second test to `weighted_mean_and_var`

This test ensures that the average and variance of samples from an
analytical funtion are close to those of the function.

gonzaponte committed:
Add third test to `weighted_mean_and_var`

This test ensured that the frequentist interpretation of the unbiased
estimator of the variance behaves correctly.

gonzaponte committed:
Add fourth test to `weighted_mean_and_var`

This test ensures the unbiased estimator of the variance when
interpreted as reliability weights behaves correctly.

gonzaponte committed:
Propagate use of new functions to the rest of the code

gonzaponte committed:
Optional implementation

This is the one exposed in:
https://github.com/nextic/IC/pull/394#issuecomment-347043372

paolafer committed:
Use size property of voxels in making tracks

jjgomezcadenas committed:
Add new normalization option "center"

[author: gonzaponte]
It is a shortcut for

```
norm_strategy = "index"
norm_opts = {"value": central_bin}
```

where central_bin is the arrays' shape tuple halved.

The idea is that the user doesn't need to know the indices
corresponding to the central bin.

[reviewer: jjgomezcadenas]
The meat of the PR is relatively straight forward and clearly written,
it's simply adding an option (center) which makes sense.

jacg committed:
402 Translate emacs init file from el-get to use-package

https://github.com/nextic/IC/pull/402

[jacg]

el-get was getting rather flaky down the years. use-package seems much
more stable.

[approved by jjgomezcadenas]

Indeed, solves #401

gonzaponte committed:
Fix double parenthesis

abotas committed:
#381 Add Zrms column to Dorothea's output

This PR includes a new column in the output DST produced by Dorothea. This column is the standard deviation of the peak in time, providing another useful variable for the analysis. It is correlated with the S2width column, but not the same. Change summary:

The kdst produced by Dorothea now contains another column (Zrms).
The Peak class now contains a new attribute (rms) and a new method (rms_above_threshold).
Cosmetics.

https://github.com/nextic/IC/pull/381

abotas committed:
extend Peak's interface

The new Peak class now has properties and methods that were frequently used in the
old pmaps Peak class.

gonzaponte committed:
Add fixtures for testing

The number of hypothesis drawings must be kept low for a good
performance.

abotas committed:
Add tests for _SensorResponses class

abotas committed:
Add tests for _Peak class

gonzaponte committed:
Add tests for PMap class

gonzaponte committed:
#405 Fix `test_weighted_mean_and_var_unbiased_frequentist`

This test was failing frequently due to the lack of compatibility
between hypothesis and flaky modules. This has ben fixed by replacing
the hypothesis dependence with a random number.

https://github.com/nextic/IC/pull/405

gonzaponte committed:
Add rms to `_Peak` interface

abotas committed:
Implement SensorResponses class

abotas committed:
Implement _Peak class

gonzaponte committed:
Implement PMap class

gonzaponte committed:
Replace "private" attributes + properties with public attributes

abotas committed:
#394 New pmaps interface

This is the first in a series of PRs that will create simpler, easier
to use pmaps for IC. This PR delineates the interface for the new pmaps.

https://github.com/nextic/IC/pull/394
see also: https://github.com/nextic/IC/issues/177

andLaing committed:
Create calib folder.

Created separate folder for 'cities' related
to calibration of sensors.

sipmPDF.py and related test moved to this folder.

jjgomezcadenas committed:
Change name of class from Sipmpdf to SipmPdf.

This is done to conform to convention (hopefully agreed by the team)
of starting class name with Capitals and then use intermediate
capitals for clarity.

andLaing committed:
Changes pythonize and make work again

Fixed error in SipmPdf class name so recognised
when capitalize used.

generalized bin/city so it can search in calib folder

changed names to be more pythonish

added sipm_pdf to command_line_execution_test.py so checked

andLaing committed:
Reverted names of program/class to approved format

Changed sipm_pdf.py -> sipmpdf.py with equivalent changes
for tests and example config files.

Updated command_line_execution_test.py to reflect the above change.

jjgomezcadenas committed:
Add functions for pedestal subtraction and normalization.

Three methods can be used to compute pedestal and return a normalized
waveform for the SiPMs.

1. Using the mean of the waveform as pedestal (fast)
2. Using a Mau to compute the pedestal (slow)
3. Returning the values above a thredhold (on top of the MAU, also
slow).

jjgomezcadenas committed:
Split peak_functions_c in two files.

peak_functions_c contained functions related with peak finding (as the
name suggests) AND functions related with sensor calibration (e.g,
removing pedestals, divinding by calibration constants). This commit
splits the original file, adding calib_sensor_functions_c, which
collects the sensor-calibration related functions. Tests files are
also split and the changes are propagated to base_cities and
peak_functions.

jjgomezcadenas committed:
Make thr a dynamic variable.

Thr (thresholds) can be used as either single threshold value or as a
vector of individual thresholds (as in the original function by
AL). There is no performance penalty here (among other things because
the function is already very slow)

jjgomezcadenas committed:
Fix bug in sipm_signal_above_trh_mau

Forgotten loop added

jjgomezcadenas committed:
Add tests.

Function sipm_signal_above_thr passes the same tests than original
function (signal_sipm)

jjgomezcadenas committed:
Add new tests.

Two new tests demonstrate that subtracting the baseline from a
waveform distributed around the baseline yields a baseline close to zero.

jjgomezcadenas committed:
Replace signal_sipm by sipm_signal_above_threshold.

The function proposed by AL uses an option (Cal) to perform three
different calculations: If Cal = 0, the function return the values of
the waveform above the baseline + MAU. If Cal = 1, the function
returns the calibrated function (wd - ped) /adc_to_pes, while if Cal=2
it returns the same but adding a MAU. The proposed alternative splits
signal_sipm in three functions, corresponding to the different values
of Cal for clarity. Replacement of signal_sipm by
sipm_signal_above_threshold in peak functions has no effect (the
results of both functions as shown by tests are identical)

jjgomezcadenas committed:
Make signal_sipm private.

The signature of signal_sipm is changed to _signal_sipm, to mark its
new status as private. The function is still used by the tests and by base_cities.

jjgomezcadenas committed:
Tests showing new functions yield same results than old function.

New tests have been added to guarantee that the new functions proposed
for the API yield the same results than the old function (now made private).

gonzaponte committed:
Add new_peak_functions module

gonzaponte committed:
Add tests for new_peak_functions

gonzaponte committed:
#413 Add function for comparing exact values

This function is intended to have similar behavior to `pytest`'s
approx function, which compares floating point numbers, but for
integers. This is particularly interesting when comparing numpy
arrays, as the expression is a bit tedious:

```
np.all(array1 == array2)
```

to be compared with

```
array1 == exactly(array2)
```

which is more pleasant to the human eye and keeps the symmetry with
floating point comparisons.

https://github.com/nextic/IC/pull/413

jjgomezcadenas committed:
403 Add SiPM spectra city

https://github.com/nextic/IC/pull/403

Addition of city and related functions for the generation of SiPM
photoelectron spectra (binned pdfs).

Additions include:

    invisible_cities/cities/sipmPDF.py (and corresponding test) which
    reads raw data and generates two types of spectra for each channel
    once converted into pe (zeroed and in pe and zeroed, in pe and
    difference from mau);

    invisible_cities/io/hist_io.py, a histogram writer to save on
    output data size.

Changes to existing functions include:

    change to invisible_cities/peak_functions_c.pyx.signal_sipm to
    include calibration option so that all bin data is saved rather than
    just selected data
        cal=0 (default) is equivalent to unchanged function,
        cal=1 saves sipmrawf-mean,
        cal=2 saves sipmrwf-mean-mau

Undid accidental include of some changes from an unrelated
branch.

Aretno committed:
Add peak XY to penthesilea hit table definition

Added Xpeak and Ypeak columns to the Hits table. Those will be the
values for the mean position of the peak (through barycenter).

Aretno committed:
Add peak XY writing to HitCollection

Added the writing of mean XY of the peaks to the HitCollection class.

gonzaponte committed:
Add new_pmaps_io_module

gonzaponte committed:
Add tests for new_pmaps_io

Aretno committed:
Adapt test to new Penthesilea structure

Add the peak XY to Penthesilea tests

Aretno committed:
Modified IO to accomodate new Penthesilea structure

Added peak XY to the hits IO functions.

gonzaponte committed:
Add useful functions to `testing_utils`

This adds:
- equatility comparisons for PMap-related classes.
- verbose names for operations of smallest possible step on floating
point numbers.

gonzaponte committed:
Adapt pmapvectors to new pmap interface

gonzaponte committed:
Adapt cities to new pmaps

gonzaponte committed:
Delete deprecated code

Aretno committed:
More tests adapted to new hDST fields.

gonzaponte committed:
Adapt filters to new pmaps

gonzaponte committed:
Add function to read run info from a file

This function is already implemented, but lives in the wrong place and
will be removed shortly.

gonzaponte committed:
Fix imports and cosmetics

gonzaponte committed:
Refactor new_pmaps_io

Using pandas' groupby is more clear and elegant.

abotas committed:
#391 Refactor interpolation in Correction class

This PR revisits the interpolation section of the Correction class. The main changes are:

Major refactoring: A chunk of code that performed interpolation has been replaced by a handy function from scipy
As a result, the linear and cubic interpolations are directly available.
The output of the __call__ method is consistent with the types of the input. Namely, if the inputs are arrays (scalars), the output contains arrays (scalars).
Tests for all of the above
Cosmetics, as usual

https://github.com/nextic/IC/pull/391

gonzaponte committed:
Refactor sipmpdf

Plus some cosmetics.

gonzaponte committed:
Add new_pmaps_functions module

gonzaponte committed:
Add tests for new_pmaps_functions

gonzaponte committed:
Revert "Refactor new_pmaps_io"

Although the version using groupby is visually more satisfying, it
introduces a mismatch between different sub-dataframes. For example:

df_A contains an event with peaks 0, 1, 2
df_B contains an event with peaks 0, 2

The previous version zips groupby objects, which would pair peak 0 in
df_A with peak 0 with df_B (ok) and peak 1 in df_A with peak 2 in
df_B. On top of introducing a mismatch between peaks, it would not run
over peak 2 in df_A.

The original version slicing dataframes manually is uglier (and
probably a bit slower for large dfs), but it is safe.

gonzaponte committed:
Move calibration functions to calib_sensors_functions

bpalmeiro committed:
Add get_chi2_and_pvalue function

Gets reduced chi2 and p-value from fit results. Previous one just took
into account the poisson-smeared case, now it is genelarized for every
smearing given by the user. Poisson case is taken by default.

bpalmeiro committed:
Replace fit with fitf for consistency

bpalmeiro committed:
Add tests for get_chi2_and_pvalue

paolafer committed:
Improved sensor simulation

fee.NOISE_DAQ reduced to remove double counting
in noise stages of PMT electronics simulation.

fee.ADC_TO_PES recalculated using faked calibration
pulses and changed to 18.71 to better calculate the
gain dispersion of the PMTs.

In sensor_functions:

Gain dispersion application changed position so that
it doesn't affect the noise which was found not to match
dispersion in noise seen in data.

charge_fluctuation function added to simulate the single
pe fluctuation using the RMS values in the database.

bpalmeiro committed:
Change fit function to meet new chi2 implementation

Be aware, now sigmas are taken as absolute errors, if given, if not,
they remain as relative weights.

bpalmeiro committed:
Add chi2-related specific tests due to changes in fit function

bpalmeiro committed:
Change and harden fit_correction test to meet new implementation

gonzaponte committed:
359 Fix chi2 and errors in fit function

https://github.com/nextic/IC/pull/359

[author: bpalmeiro]

- Chi2
Implemented chi2 computation, based on scipy.stat.chisquare is only
valid when data smearing is Poisson distributed. So, a general chi2
function is added, with formula extracted from [1].

- Errors
Fit errors, given by scipy.optimize.curve_fit were not correct as flag
absolute_sigma was set to false. When true, "[data] sigma describes
one standard deviation errors of the input data points", so covariance
matrix estimation is based on these values. Otherwise, data sigma is
only taken as "relative weights of the data points" and covariance
matrix is not affected by the overall magnitude of the values in
sigma, just by relative magnitudes.

Both improvements are shown in
https://github.com/bpalmeiro/ICARO/blob/Chi2_demostration/icaro/Chi2_demonstration/Chi2_demonstration.ipynb

[1] Numerical Recipes. The Art of Scientific Computing, 3rd Edition,
2007, ISBN 0-521-88068-8. Pages 778-780

[reviewer: gonzaponte]

This PR fixes a long-term issue with the definition of the chi2 and
the p-value. The calculations have been encapsulated within a
function, which is tested. The previous test suite for the rest of the
code has also been fixed, improved and extended.

bpalmeiro committed:
Use already defined poisson error function

bpalmeiro committed:
Remove non zero selection when chi2 is computed

It is not currently necessary due to the changes in PR #359

bpalmeiro committed:
Add a test that demonstrates that invalid sigma values are accepted

bpalmeiro committed:
Add safety check when sigma negative or zero values are given

paolafer committed:
Change default value for strictness in voxelize_hits

paolafer committed:
Test voxel size in both strictness modes

gonzaponte committed:
Refactor calib_sensors_functions

And rename calib_sensor_functions to calib_sensors_functions (notice
the trailing "s" in sensors).

gonzaponte committed:
Refactor calib_sensors_functions_test

gonzaponte committed:
Introduce ICDATADIR fixture

This reduces the noise when using test files.

gonzaponte committed:
Prepare fixtures and data files for testing

gonzaponte committed:
Remove new_ from file names

gonzaponte committed:
Remove new_ from imports

paolafer committed:
Use strict voxel size for test about respecting voxel dimensions

paolafer committed:
424 Eliminate misleading dependence on voxel size from functions

[author: paolafer]

Eliminate explicit dependence of make_track_graph function on voxel
size, since now voxels know their size.

[reviewer: jjgomezcadenas]

gonzaponte committed:
Rename misspelled variables

gonzaponte committed:
Add file with example rwfs and the corresponding blr wfs

gonzaponte committed:
Add tests for blr

jahernando committed:
421 Small fixes in fit functions

https://github.com/nextic/IC/pull/421

[author: bpalmeiro]

These are just small changes that remained from #359:
- Remove a non-zero selection in fit function, no longer necessary.
- Add a test that demonstrates that invalid values for sigma were
accepted.
- Correct correspondingly the code for checking invalid values in
sigma.

[reviewer: jahernando]

The PR solves an issue related with the treatment of the data when the
user sends null errors to the fit function. Now it raises an
exception.

jacg committed:
406 Implement new PMAPs interface

https://github.com/nextic/IC/pull/406

[authors: abotas & gonzaponte]

This PR consists of a full parallel python implementation with a
corresponding test suite for the new PMAPs following their interface
defined in PR #394.
This is the second step defined in #407.

[reviewer: jacg]

Clear, well-tested and desperately needed.

gonzaponte committed:
Move `poisson_sigma` to its own module

This solves an import loop created in #421.

gonzaponte committed:
Fix poisson_sigma imports

gonzaponte committed:
Cosmetics

gonzaponte committed:
Add tests for stat_functions

gonzaponte committed:
428 Remove flaky report if tests succeed

The number of flaky tests has recently been increased and the report
introduces too much noise to `pytest`'s output. This commit introduces
a flag to pytest that removes the report if the tests pass.

jjgomezcadenas committed:
427 Add stat functions module

https://github.com/nextic/IC/pull/427

[author: gonzaponte]

This PR solves an import loop introduced in #421.

The function `poisson_sigma` has been moved from
`icaro/hst_functions.py` to its own module
`core/stat_functions.py`. The function is now tested.

[reviewer: jjgomezcadenas]

paolafer committed:
Take masked channel into account in the corona algorithm

jjgomezcadenas committed:
423 Switchable voxel size strictness

https://github.com/nextic/IC/pull/423

[author: paolafer]

Originally, `voxelize_hits` created voxels slightly smaller than the
required voxel size.  We have added the option of using voxels of
strict size. Fixes #396.

[reviewer: jjgomezcadenas]

Simple and clear.

gonzaponte committed:
Refactor hypothesis generators

This produces more realistic data

jjgomezcadenas committed:
425 Add tests for the BLR algorithm

https://github.com/nextic/IC/pull/425

[author: gonzaponte]

This PR introduces some tests to the BLR algorithm, which belongs to
the core of our processing chain. There are two types of tests:

- Property checking: by using a toy signal, we check that the integral
is positive and that the baseline is recovered.

- Stability: by using some real waveforms we store the current value of
the blr waveform in order to have something to compare with in the
future. This assumes we trust the current algorithm.

[reviewer: jjgomezcadenas]

There was plenty of tests and studies for BLR but they never made it to IC.

jjgomezcadenas committed:
417 Adapt IC to new PMAPs model

https://github.com/nextic/IC/pull/417

[author: gonzaponte & abotas]
This PR adapts all existing code in IC to the new PMap interface. This
includes a big refactor of peak_functions, pmap_functions and
calib_sensor_functions, which have been fully de-cythonized. The IO
functions have also been rewritten to match the new representation.

Everything is thoroughly tested.

[reviewers: jjgomezcadenas & jacg]

gonzaponte committed:
Fix mismatch between database and input wfs

gonzaponte committed:
Use class attributes in methods rather than passing them around

gonzaponte committed:
Use ICDATADIR fixture in diomira tests

gonzaponte committed:
Add test to demonstrate mismatch between database and input wfs

gonzaponte committed:
Add safety check to diomira

This ensures that the trigger channel is not masked.

gonzaponte committed:
Add test triggering on masked channels

This demonstrates the lack of protection against it.

gonzaponte committed:
Move calculation of trigger channels to initialization

It is constant and therefore doesn't need to be calculated for each
event.

gonzaponte committed:
Do not apply blr algorithm to unused wfs

Saves precious time

Aretno committed:
Adapt HitCity to new Penthesilea structure

Added a peak xy_position function with standard KrCity parameters.
Use that function to obtain the peak position as it is done in
Dorothea and add that information to the Hit.

Aretno committed:
Cosmetics

gonzaponte committed:
415 Add mean XY of the peak to hDST

https://github.com/nextic/IC/pull/415

[author: Aretno]
This PR adds the Xpeak and Ypeak columns to Penthesilea hits
output (hDST). These are the XY mean position of the peak and is added
to the Hit class as an XY object.

For calculating the position a function called
`compute_xy_peak_position` has been introduced. This function is
almost equal to `compute_xy_position` from the kDSTs with the
exception that the configuration for corona is hardcoded to be the
general configuration (to avoid using the corona config used for the
hit-building). In case it fails, the mean position is set to be 0.,
0., as with NN hits.

To make the code backwards compatible the `getattr` built-in function
is used, in case the hDST does not have the XYpeak fields, the hits
will be loaded using a default value of -1000 for those fields.

[reviewer: gonzaponte]
This PR extends Penthesilea to include peak magnitudes to the
output. The test suite has been accommodated to check the values of
the new variables. Reviewed with peras.

gonzaponte committed:
Add safety check to irene

This addresses the bug demonstrated in the last commit

gonzaponte committed:
Add safety check to PCity

To maintain backward compatibility, a safety check is introduced. The
counter implemented keeps track of the amount of events with this
problem.

gonzaponte committed:
Add test to demonstrate bug in irene

In irene, the peak search can yield zero peaks that match the features
provided in the configuration file. In such cases, the event is not
filtered out. As a result, the output file contains those event
numbers in the run info array, but not in the pmap-related tables.

gonzaponte committed:
Add tests to demonstrate crash in PCity

Due to the bug demonstrated in
8677e317f095ec9368c5595e99511988f6211204, some files contain a
mismatch between the event number table and the pmap tables. This
commit demonstrates that an error is raised in those cases.

gonzaponte committed:
Fix test_irene_runinfo_run_2983

jjgomezcadenas committed:
432 Cities safety checks

[author: @"gonzaponte]

This PR fixes a bug in the output of Irene, where the event number of
empty (at the pmap level) events were being stored, creating an
inconsistency between two different tables in the file. At the same
time, PCity contained a check when reading the input that allowed the
bug to pass silently. In #417 this check was removed, which revealed
the bug.

This PR filters out empty events from Irene's output and reintroduces
the check in PCity for backward-compatibility (so we can run the city
on old data). Three tests have been introduced to ensure this bug
doesn't cause any further problems.

[approved by: @jjgomezcadenas]

Good catch, and excellent fix.

Commit:
Adds functions to read HDF5 Nexus output and corresponding tests

paolafer committed:
Add function to read waveforms and relative test

Commit:
Replaces test file and uses pytables instead of h5py

Commit:
Fixes formatting issues and removes test filename fixture parameters

Commit:
Decodes strings from Nexus test file; only saves "ACTIVE" hits

paolafer committed:
Replace tracks with info for more clarity

Commit:
Sets electron_MCRD_file as a fixture parameter again

paolafer committed:
Update MCParticle and MCHit classes

andLaing committed:
445 Fix to potential error in NoiseSampler

https://github.com/nextic/IC/pull/445

[author: andLaing]

Code added baseline to samples before conversion to ADC which is an
error in the case of a baseline != 0.

Changed class to have access to sipm gains so that conversion could be
done in function.

Affected functions and tests altered to take change into account.

[reviewer: gonzaponte]

I would like to have a test that demonstrates the bug, but since it
depends on the database I reckon it will be too difficult to do
now. However, as soon as #420 is implemented, we should dedicate some
time to pay this debt.

I am happy with this PR otherwise. It introduces a bit of an overhead
as it includes another call to the database, which will be addressed
shortly.

jahernando committed:
438 Add functions to read HDF5 Nexus output and corresponding tests

https://github.com/nextic/IC/pull/438

[author: jerenner]

These functions read MC hits and particles from Nexus HDF5
output. They and their tests are similar to the existing functions for
reading MC hits.

[reviewer: jahernando]

The PR is approved, subjected to refactoring.

paolafer committed:
Change MC true structure to new hdf5 format

paolafer committed:
Add test for corona with masked channels

paolafer committed:
Pass masked sipms in base_cities

paolafer committed:
Add tests for new functions in xy_algorithms

gonzaponte committed:
430 Take into account masked channels in corona

https://github.com/nextic/IC/pull/430

[author: paolafer]

This PR addresses issue #387, making possible to take into account the
presence of masked SiPMs in the corona algorithm, when calculating the
minimum number of SiPMs within the radius.  Specifically, if N masked
SiPMs are found, the minimum number of SiPMs required becomes msipm -
N.

[reviewer: gonzaponte]

Extension to the corona algorithm to take into account the masked
channels. Everything is tested.

gonzaponte committed:
443 Allow database file selection via parameter

https://github.com/nextic/IC/pull/447

[author: gonzaponte]

This makes testing easier as it allows the use a different, maybe
custom, database in the test suite.

[reviewer: jmbenlloch]

This is a simple change to ease the future tests involving the
database. I have already rebased this PR and therefore I'll approve
and merge.

paolafer committed:
Modify MC test files and relative code with new format

paolafer committed:
Fix way of writing particles and hits tables

paolafer committed:
Remove casting, not needed with the new file

paolafer committed:
Add new files and adapt tests to them and the new MC format

paolafer committed:
Make tests that don't use MC think that file is data

paolafer committed:
Update group name of table and mark skip for now

paolafer committed:
Update files using the latest version of the writer

paolafer committed:
Style improvements

paolafer committed:
Make search for event faster in evt by evt reader

paolafer committed:
Re-use read_mcinfo_evt_by_evt function in the all events reader

paolafer committed:
Add type hint

paolafer committed:
Fix bug in filtered event writing and add test

paolafer committed:
Cosmetics

paolafer committed:
Add test to hit and particle content in filtered events

paolafer committed:
Add test for specific cases: skip first evt, first evt well written

paolafer committed:
Modify parameter according to the window of the new file

paolafer committed:
Add new test file and slightly tune test parameter

paolafer committed:
Fix name of group in sensor data and relative test

paolafer committed:
Add Waveform class to deal with sensor response

paolafer committed:
Fix retrieval of time binning from sensors

jmbenlloch committed:
450 Take the proper interval from database

paolafer committed:
Put nexus tag name in file and add nexus original file

paolafer committed:
Adapt tests to new files

paolafer committed:
Use NamedTuples

paolafer committed:
Update test file name in configuration file

jmbenlloch committed:
431 Fix diomira masked channels

https://github.com/nextic/IC/pull/431

[author: gonzaponte]

This PR:

    Demonstrates that Diomira doesn't work with masked pmts unless the
    input is already masked and fixes it.
    Demonstrates that there isn't any protection at all against
    triggering on masked channels. Now it raises an error.
    Speeds up Diomira a bit by not deconvoluting all waveforms (only
    those needed for trigger simulation).

[reviewer: paolafer]

This PR addresses a bug in the dealing of masked PMTs in Monte Carlo
simulation. The solution proposed works and a test has been added
which demonstrates the need for this. Also, a protection against
triggering on masked PMTs has been added. I approve it!

paolafer committed:
Add exception for sensor time binning not known

jjgomezcadenas committed:
Add new functions to core

These functions are useful and of general purpose

jjgomezcadenas committed:
Add tests for new functions in core

gonzaponte committed:
Add function to compute poisson pmf

The scipy implementation is significatively slower.

jjgomezcadenas committed:
Add utility to ensure all elements in an array are close

andLaing committed:
Fix typo

jjgomezcadenas committed:
Add waveform generator

This module is used to generate toy waveforms (square and sawteeh) for
testing and tutorials purposes.

andLaing committed:
Add new functions for calibration

andLaing committed:
Add tests for new calibration functions

andLaing committed:
Extend calib_sensors_functions with mode and remove mau

It also includes a refactor and some documentation.

Calibration with mau has been removed due to lack of accuracy.

andLaing committed:
Add new tests to calib_sensors_functions

It also includes some refactoring

andLaing committed:
Add functions to fit spe data

andLaing committed:
Add tests for spe_response

gonzaponte committed:
Refactor sipmpdf

andLaing committed:
Add pmt gain city

andLaing committed:
Add sipm gain city

andLaing committed:
Use mode for sipm calibration

andLaing committed:
Add new cities to command line execution test

gonzaponte committed:
440 Add sensor calibration cities

https://github.com/nextic/IC/pull/440

[author: andLaing]

New cities and functions for the calibration of the pmts and sipms.
1. New cities: pmtgain and sipmgain
2. Functions common to all or some calibration cities centralised in
calib_functions
3. New functions to fit spe data added to reco relevant for extracting
calibration parameters.

[reviewer: gonzaponte]

Implementation of the necessary functions for calibration and the
corresponding cities. It also includes a refactor of
calib_sensor_functions, adding the mode for baseline calculation and
removing the mau. Some utilities have also been added.

Great work.

andLaing committed:
Added covariance matrix to FitFunction

FitFunction tuple now contains the covariance
matrix of the fit to ease error propagation.

covariance matrix added in fit_functions.fit

jmbenlloch committed:
449 Read and write MC true info

https://github.com/nextic/IC/pull/449

[author: paolafer]

In this PR the writing and reading code of MC true information has
been updated to the new hdf5 format generated by nexus.

Three tables are now stored: one with the particle information, one
with the hit information and a third one with the information needed
to relate the first two tables to each event. These three tables are
read and written again and the writing takes into account the possible
filtering of some events.

Some of the MC test files have been generated again, to have the new
hdf5 format, and some of them have been deleted because not in use
anymore.

[reviewer: gonzaponte]

Rearrangement of the mc info reader to accommodate new data structure
produced by the simulation software. The test suite has also been
adapted and extended, so everything is quite well tested.

Great work.

gonzaponte committed:
Move shift_to_bin_centers function to core

gonzaponte committed:
Replace profile functions with a faster implementation

This also follows the principle of not reinventing the wheel
by using standard functions.

jjgomezcadenas committed:
456 Replace implementation of profile functions with faster ones

https://github.com/nextic/IC/pull/456

[author: gonzaponte]

In this PR, the functions `profile*` are replaced with a faster
implementation. The new implementations use standard functions, which
follows the principle of not reinventing the wheel.

However, for simpler cases there is a better set of functions in
`scipy.stats`: `binned_statistics`, `binned_statistics_2d` and
`binned_statistics_dd`. Those functions are designed to do exactly what
we are want to do: bin data on some axes and apply operations on the
remaining one. But there is one caveat: they can only return one
statistic per bin. We need to get 3 in our case: mean, std and count,
which would imply three calls to those functions, worsening the
overall performance of the function. Thus, I have explored a
implementation using `pandas`' dataframes, which is a factor 2-3 faster
than our original implementation (in 1-d) and a factor 3-4 faster than
the implementation using `scipy` functions. Those speed factors increase
very rapidly with the number of dimensions of the original dataset.

[reviewer: jjgomezcadenas]

This is a very smart adaption of an existing resource to our case,
avoiding the re-invetion of the wheel and saving time.

gonzaponte committed:
Add test to ensure return values are correct

There is a subtle difference between the computation of the std
deviation in `numpy` and in `pandas`. The standard deviation is
defined as:

```
std = sum((x - x.mean())**2)/(N - ddof)
```

For `numpy`, ddof == 0 and for `pandas` ddof == 1, which makes a
subtle difference for large datasets and a moderate one for smaller
ones.

This ensures the output is exactly what we expect in a simple case.
As stated in the documentation of `np.std`:

> In standard statistical practice, ddof=1 provides an unbiased
> estimator of the variance of the infinite population. ddof=0
> provides a maximum likelihood estimate of the variance for normally
> distributed variables.

Thus, the correct version of `std` for our intentions is with ddof=1,
which is the default in `pandas`, but not in `numpy`.

gonzaponte committed:
Add test to demonstrate bug when drop_nan is False

The output of the functions when drop_nan is False is wrongly placed
in the output array. This test demonstrates it.

gonzaponte committed:
Fix bug in the output values when drop_nan is False

andLaing committed:
Added test for covariance matrix

Added a test which checks the shape of the
covariance matrix for a Gaussian fit.

andLaing committed:
Implementation of reviewers requests

Tidied up pythonification

Moved check of zeros in SiPM database 1pe Sigmas
to the database function DataSiPM.

paolafer committed:
Add test to show failing with multiple files as input

paolafer committed:
Fix problem when reading more than one file in the same job

paolafer committed:
Add initialization of a fourth variable

Commit:
Fixed missing gain_mau option in pmtgain

Re-added function for the second option which
had been removed accidentally in previous
PR review.

paolafer committed:
Change types for tables

paolafer committed:
Adapt code to change in types

jjgomezcadenas committed:
457 Patch 456: fix standard deviation and bug with `drop_nan=False`

https://github.com/nextic/IC/pull/457

[author: gonzaponte]

This is a patch to #456. The test suite didn't catch two errors in the
previous PR:

- The definition of the standard deviation changed, but it was subtle,
particularly for large datasets (See the description of 97c2423).

- When the value of the keyword-argument `drop_nan` is False, the
output arrays had their values shifted one position to the right.

The test suite has been extended to ensure these two bugs don't show
up again. Also, the current definition of the standard deviation is
the right one.

[reviewer: jjgomezcadenas]

A nice catch, nicely demonstrated and fixed.

jjgomezcadenas committed:
459 Reset MC writer attributes when reading a new file in Irene

https://github.com/nextic/IC/pull/459

[author: paolafer]

This PR fixes an issue that happens when processing multiple input
files with irene city. First, a test has been added to catch the
problem, then a fix has been implemented.

[reviewer: jjgomezcadenas]

This is a necessary and well documented fix and it includes a correct
test.

Commit:
Added tests for the non default processing calibration options

Added the test for gain_mau in pmtgain_test

Added test for subtract_median in sipmgain_test

Commit:
Added additional basic test to covariance matrix

test_covariance_matrix_values

Commit:
Cosmetics

gonzaponte committed:
452 Add covariance matrix to FitFunction's output

https://github.com/nextic/IC/pull/452

[author: andLaing]

FitFunction namedtuple now contains the covariance matrix of the fit
to ease error propagation.

[reviewer: gonzaponte]

Nice addition to the output of the fit function. It also adds a couple
of tests to ensure the output is correct.

gonzaponte committed:
460 Fix missing `gain_mau` option in pmtgain

https://github.com/nextic/IC/pull/460

[author: andLaing]

Re-added function for the second option which had been removed
accidentally in previous review of #440.

[reviewer: gonzaponte]

Perfect. Fixes functionality + extends test suite to avoid this kind
of mistakes in the future.

gonzaponte committed:
454 Improve sensor simulation

https://github.com/nextic/IC/pull/454

[author: andLaing]

This pull request combines a review of the existing sensor simulation
functions and constants made with the help of @jjgomezcadenas and
V. Herrero and work to include the charge fluctuation to the sensor
simulation carried out by myself and @paolafer .

The PMT low frequency noise simulation has been removed for now since
its proper inclusion requires changes to the database.

Change summary:
- fee.NOISE_DAQ reduced to remove double counting in noise stages of
PMT electronics simulation.
- fee.ADC_TO_PES recalculated using faked calibration pulses and changed
to 18.71 to better calculate the gain dispersion of the PMTs.
- Gain dispersion application changed position so that it doesn't
affect the noise which was found not to match dispersion in noise seen
in data.
- `charge_fluctuation` function added to simulate the single pe
fluctuation using the RMS values in the database.

[reviewer: gonzaponte]

This PR adds charge fluctuation to the simulation of the sensors'
response. The database functions have been modified to provide the
relevant values for this task. Finally, some fee variables have been
updated to better match the data.

paolafer committed:
Unify type of event number across IC

paolafer committed:
Trick to reuse the same test files with the type change

jmbenlloch committed:
462  Change types in MC true info

https://github.com/nextic/IC/pull/462

[author: paolafer]

This PR is the adaption of the IC code to a problem of overflow in
nexus MC true tables.
The types of the columns of the extents table have been changed from
unsigned int or int to uint64 or int64 in order to be able to
represent a larger number of events. Strangely (to me), it turns out
that, when adding a uint64 to an int, python gives as a result a
float64 (see, for instance, this thread). Therefore, in this PR I
simply cast my uint64 variable to int, which, if I'm not wrong, is the
same as a long in python, so it should be ok.

[reviewer: jmbenlloch]

I think this update is ok. It makes uniform the type of event numbers
across the different tables and updates the table MC/extents to avoid
overflows. We need this update to be able to continue with the
development of related projects like Nexus. I have rebased this branch
to the latest master. If all tests pass I will approve and merge.

gonzaponte committed:
Add tests to demontrate bugs

The sensors' waveforms are not properly matched to their IDs because
the IDs are sorted (independently from the waveforms) when read. This
is illustrated in these tests.

gonzaponte committed:
Fix bug in `build_*_responses`

Besides removing duplicates, `np.unique` sorts the input values, which
creates a mismatch between the sensor's id and its
waveform. `pd.unique` on the other hand preserves order with one
caveat: it takes into consideration only *consecutive*
duplicates (i.e. repeated values with a different one in between are
counted multiple times). Since our waveforms are always written on a
sensor-by-sensor basis, the same ID cannot be read twice in the same
peak and this tool is still useful. Two alternative
implementations (one with `np.unique` and one with
`itertools.groupby`) have also been tested and the `pd.unique` is the
fastest one, particularly for long arrays.

gonzaponte committed:
Add fix using `itertools.groupby`

gonzaponte committed:
Archive itertools implementation

gonzaponte committed:
Add fix using `numpy.unique`

gonzaponte committed:
Archive numpy implementation

Commit:
468 Fix pmap io

https://github.com/nextic/IC/pull/468

[author: gonzaponte]

The SiPMs waveforms were not properly matched to each sensor. This was
caused by `np.unique` which, besides removing duplicates, sorts the
elements. This can also be a problem in the PMTs waveforms, although
the PMTs are, in principle, always sorted.

A couple of tests have been added to demonstrate the situation and to
ensure this doesn't happen again in the future.

The solution consists of replacing `np.unique` by `pd.unique`, which
preserves the original order. There is one caveat, though: `pd.unique`
removes only consecutive duplicate elements. This is ok as our data is
stored on a sensor-by-sensor basis and there should not be
non-consecutive duplicate elements.

Two other alternative implementations have been also considered, but
the one using `pd.unique` is the fastest.

[reviewer: jmbenlloch]

This is a VERY important bug fix. The solution chosen by @gonzaponte
is the fastest one and also includes two simple tests to avoid this
situation in the future.
